### PR TITLE
Fix orphan pkg check algorithm

### PR DESCRIPTION
**Pull request by asdil12**
https://bugzilla.suse.com/show_bug.cgi?id=1198940

### CODE REVIEW

**Comment by asdil12** [comment]
openqa-clone-custom-git-refspec https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/14828 https://openqa.suse.de/tests/8685554 ZYPPER_WHITELISTED_ORPHANS= SCHEDULE=tests/jeos/firstboot,tests/console/suseconnect_scc,tests/console/orphaned_packages_check
Github oauth token provided, performing authenticated requests
Created job #8689014: sle-15-SP4-JeOS-for-RaspberryPi-aarch64-Build1.138-jeos-main-RPi@aarch64 -> https://openqa.suse.de/t8689014

**Comment by asdil12** [comment]
openqa-clone-custom-git-refspec https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/14828 https://openqa.suse.de/tests/8685554 ZYPPER_WHITELISTED_ORPHANS= SCHEDULE=tests/jeos/firstrun,tests/console/suseconnect_scc,tests/console/orphaned_packages_check
Github oauth token provided, performing authenticated requests
Created job #8689025: sle-15-SP4-JeOS-for-RaspberryPi-aarch64-Build1.138-jeos-main-RPi@aarch64 -> https://openqa.suse.de/t8689025

**Comment by asdil12** [comment]
openqa-clone-custom-git-refspec https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/14828 https://openqa.suse.de/tests/8685554 ZYPPER_WHITELISTED_ORPHANS= SCHEDULE=tests/jeos/firstrun,tests/console/suseconnect_scc,tests/console/orphaned_packages_check
Github oauth token provided, performing authenticated requests
Created job #8689062: sle-15-SP4-JeOS-for-RaspberryPi-aarch64-Build1.138-jeos-main-RPi@aarch64 -> https://openqa.suse.de/t8689062

**Comment by mloviska** [comment]
IMO, the registration should be done outside this module, but that might be a matter of taste. Otherwise LGTM

File: tests/console/orphaned_packages_check.pm [hunk 1]
```diff
@@ -59,7 +60,12 @@ sub run {
 
     record_info((is_offline_upgrade_or_livecd) ? 'Upgrade/LiveCD' : 'No upgrade/LiveCD', 'Upgraded or installed from LIVECD can possibly cause orphans');
 
-    # Orphans are also expected in JeOS without SDK module (jeos-firstboot, jeos-license and live-langset-data)
+    if (is_sle() && is_jeos()) {
+        # Orphans are also expected in JeOS without SDK module (jeos-firstboot, jeos-license and live-langset-data)
```
Comment (Vogtinator) [inline tests/console/orphaned_packages_check.pm:64]: While you're already touching this:

```suggestion
        # Orphans are also expected in JeOS without SDK module (jeos-firstboot, jeos-licenses and live-langset-data)
```

**Review by Vogtinator — APPROVED** [review]

**Comment by Vogtinator** [comment]
> IMO, the registration should be done outside this module, but that might be a matter of taste. Otherwise LGTM

Agreed, but not necessarily only a matter of taste, as this also influences later test modules. I don't think it's that different from other modules though, which also perform registration on demand.

**Comment by asdil12** [comment]
The only remaining question is why my verification run (https://openqa.suse.de/tests/8689181#step/orphaned_packages_check/21) doesn't fail as it should due to `jeos-firstboot-rpiwifi` missing in the dev repo. I disabled the whitelist for this VR, so it should actually fail. 
But it just reports: `Missing: None`

**Comment by Vogtinator** [comment]
> The only remaining question is why my verification run (https://openqa.suse.de/tests/8689181#step/orphaned_packages_check/21) doesn't fail as it should due to `jeos-firstboot-rpiwifi` missing in the dev repo. I disabled the whitelist for this VR, so it should actually fail. But it just reports: `Missing: None`

That's a good question...

**Comment by mloviska** [comment]
> The only remaining question is why my verification run (https://openqa.suse.de/tests/8689181#step/orphaned_packages_check/21) doesn't fail as it should due to `jeos-firstboot-rpiwifi` missing in the dev repo. I disabled the whitelist for this VR, so it should actually fail. But it just reports: `Missing: None`

I see that problem, it is because the `@missed_orphans` is empty as it is in the whitelist branch. I can fix that after merging your PR.

**Comment by mloviska** [comment]
> > The only remaining question is why my verification run (https://openqa.suse.de/tests/8689181#step/orphaned_packages_check/21) doesn't fail as it should due to `jeos-firstboot-rpiwifi` missing in the dev repo. I disabled the whitelist for this VR, so it should actually fail. But it just reports: `Missing: None`
> 
> That's a good question...

Does not work here as expected `my @missed_orphans = grep(s/\s+//g, @{$args{zypper_orphans}});`

Apparently it should be `@missed_orphans = map { s/^\s+|\s+$//gr } @{$args{zypper_orphans};`

**Comment by mloviska** [comment]
```
my @list = (" test ", "test2 ", "   test3");
my @missed_orphans = map { s/^\s+|\s+$//gr } @list;
print "TEST1 ... " . Dumper(@missed_orphans);
print "\n";

my @list = ("  test4   ");
@missed_orphans = map { s/^\s+|\s+$//gr } @list;
print "TEST2 ... " . Dumper(@missed_orphans);
print "\n";
```

yields

```
TEST1 ... $VAR1 = 'test';
$VAR2 = 'test2';
$VAR3 = 'test3';

TEST2 ... $VAR1 = 'test4';
```

**Comment by mloviska** [comment]
After adding `/r` and replacing `grep` for `map` it should work.

```
my @list = ("  test4   ");
@missed_orphans = grep(s/^\s+|\s+$//gr, @list);
print "TEST4 ... " . Dumper(@missed_orphans);

my @list = ("test4.1");
@missed_orphans = grep(s/^\s+|\s+$//gr, @list);
print "TEST4.1 ... " . Dumper(@missed_orphans);

my @list = ("test4.2");
@missed_orphans = map {s/^\s+|\s+$//gr} @list;
print "TEST4.2 ... " . Dumper(@missed_orphans);

my @list = ("     test4.3 ");
@missed_orphans = map {s/^\s+|\s+$//gr} @list;
print "TEST4.3 ... " . Dumper(@missed_orphans);
```

```
TEST4 ... $VAR1 = '  test4   ';
TEST4.1 ... $VAR1 = 'test4.1';
TEST4.2 ... $VAR1 = 'test4.2';
TEST4.3 ... $VAR1 = 'test4.3';
```

**Comment by asdil12** [comment]
https://openqa.suse.de/tests/8689820

**Comment by asdil12** [comment]
https://openqa.suse.de/tests/8689838

File: tests/console/orphaned_packages_check.pm [hunk 2]
```diff
@@ -59,7 +60,12 @@ sub run {
 
     record_info((is_offline_upgrade_or_livecd) ? 'Upgrade/LiveCD' : 'No upgrade/LiveCD', 'Upgraded or installed from LIVECD can possibly cause orphans');
 
-    # Orphans are also expected in JeOS without SDK module (jeos-firstboot, jeos-license and live-langset-data)
+    if (is_sle() && is_jeos()) {
```
Comment (mloviska) [inline tests/console/orphaned_packages_check.pm:63]: My fault that I haven't mentioned this earlier! :facepalm: 

You would have achieved the same with `SCC_ADDONS=base,serverapp,desktop,sdk` as you need to use `console/suseconnect_scc.pm` to register the system. Sorry!

**Pull request merged**